stages/org.osbuild.zipl.inst

```diff
@@ -74,8 +74,16 @@ def find_kernel(root, kernel: str):
         if version != kernel:
             continue
 
-        linux = root + params["linux"]
-        initrd = root + params["initrd"]
+        # Find the prefix that needs to be added to get the full path to
+        # the kernel/initrd. If /boot is a separate filesystem then the
+        # path inside the blscfg won't have that prefix on it and we need
+        # to add it here.
+        prefix = root
+        if os.path.ismount(os.path.join(root, "boot")):
+            prefix = os.path.join(root, "boot")
```

Comment:
Can you please elaborate on the problem that the current state causes? We do use separate mountpoint for `/boot` on RHEL image for s390x and I'm trying to understand what is the impact. Because AFAIK, we didn't see any complains about it yet :thinking: 

In a different note, I'm wondering if this should not be explicit, instead of implicit. IOW, the tool that generates the manifest should know if the `/boot` is on a separate partition and the stage should not need to inspect mountpoints...

Reply:
> Can you please elaborate on the problem that the current state causes? We do use separate mountpoint for `/boot` on RHEL image for s390x and I'm trying to understand what is the impact. Because AFAIK, we didn't see any complains about it yet 🤔

Any chance you could share a link to a json input manifest for the RHEL s390x image? We might be able to learn from it.

Reply:
You can check e.g. https://github.com/osbuild/osbuild-composer/blob/main/test/data/manifests/centos_9-s390x-qcow2-boot.json

Reply:
ok. I think the RHEL config is "working" without this PR because it [is using](https://github.com/osbuild/osbuild-composer/blob/0c0a7586002c4c059819deaff02b52e0c8891678/test/data/manifests/centos_9-s390x-qcow2-boot.json#L5323-L5326) the [fix-bls](https://github.com/osbuild/osbuild/blob/main/stages/org.osbuild.fix-bls) stage, which honestly I suspect was written and used as a workaround for a couple of issues [1](https://github.com/osbuild/osbuild-composer/raw/main/test/data/manifests/centos_9-s390x-qcow2-boot.json) [2](https://github.com/osbuild/osbuild/issues/1495) that were recently fixed.

I don't really think the BLS config needs fixing here because (at least on a grub system) the prefix (path) to the kernel/initrd are relative to the root of that filesystem and since boot is on a separate filesystem then to find the path we'd need to prepend `/boot`. Here is a current Fedora CoreOS system today (not built with osbuild)

```
core@coreos-s390x-builder:~$ sudo cat /boot/loader/entries/ostree-1-fedora-coreos.conf | grep linux
linux /ostree/fedora-coreos-78f7d269565a6139e8969ea8b23f981ce4c6019afa5a4564a51d5395cf25afa0/vmlinuz-6.6.8-200.fc39.s390x
core@coreos-s390x-builder:~$ 
core@coreos-s390x-builder:~$ ls /boot/ostree/fedora-coreos-78f7d269565a6139e8969ea8b23f981ce4c6019afa5a4564a51d5395cf25afa0/vmlinuz-6.6.8-200.fc39.s390x 
/boot/ostree/fedora-coreos-78f7d269565a6139e8969ea8b23f981ce4c6019afa5a4564a51d5395cf25afa0/vmlinuz-6.6.8-200.fc39.s390x
```

Reply:
The disk doesn't exist yet when we're building the image and fixing the bls.  That's fundamentally different to how we build images.  We prepare everything ahead of time and only copy things to disk at the end, primarily for efficiency but also for reusability of the OS pipelines.

One thing I'm not happy about here is (again) the stage detecting things about the tree or the mounts.  If a `prefix` is needed for some of the options and the stage has a device with mounts attached to it, why does the prefix need to be detected?  The value should be known when creating the manifest and the paths can be set accordingly in the stage options, can't they?

Reply:
> The disk doesn't exist yet when we're building the image and fixing the bls. That's fundamentally different to how we build images. We prepare everything ahead of time and only copy things to disk at the end, primarily for efficiency but also for reusability of the OS pipelines.

OK I decided to dig in a little bit more because I realized I didn't 100% understand the `fix-bls` stage. Armed with `--break=org.osbuild.fix-bls` I did a local build of the [corresponding x86 image](https://github.com/osbuild/osbuild-composer/blob/e289b763e7bbadb7a95e566d9d365efb79d47557/test/data/manifests/centos_9-x86_64-qcow2-boot.json)


Here is the `fix-bls` config:

```
          {
            "type": "org.osbuild.fix-bls",
            "options": {
              "prefix": ""
            }
          },
```

and here is what I see before `fix-bls` runs:

```
bash-5.1# cat /run/osbuild/tree/boot/loader/entries/ffffffffffffffffffffffffffffffff-5.14.0-214.el9.x86_64.conf
title CentOS Stream (5.14.0-214.el9.x86_64) 9
version 5.14.0-214.el9.x86_64
linux /root/var/osbuild/store/stage/uuid-e412a9ca6de34867a04f812cd0dcd623/data/tree/boot/vmlinuz-5.14.0-214.el9.x86_64
initrd /root/var/osbuild/store/stage/uuid-e412a9ca6de34867a04f812cd0dcd623/data/tree/boot/initramfs-5.14.0-214.el9.x86_64.img $tuned_initrd
options root=UUID=6e4ff95f-f662-45ee-a82a-bdf44a2d0b75 console=tty0 console=ttyS0,115200n8 no_timer_check net.ifnames=0 crashkernel=1G-4G:192M,4G-64G:256M,64G-:512M $tuned_params
grub_users $grub_users
grub_arg --unrestricted
grub_class centos
```

and after:

```
bash-5.1# cat /run/osbuild/tree/boot/loader/entries/ffffffffffffffffffffffffffffffff-5.14.0-214.el9.x86_64.conf 
title CentOS Stream (5.14.0-214.el9.x86_64) 9
version 5.14.0-214.el9.x86_64
linux /vmlinuz-5.14.0-214.el9.x86_64
initrd /initramfs-5.14.0-214.el9.x86_64.img $tuned_initrd
options root=UUID=6e4ff95f-f662-45ee-a82a-bdf44a2d0b75 console=tty0 console=ttyS0,115200n8 no_timer_check net.ifnames=0 crashkernel=1G-4G:192M,4G-64G:256M,64G-:512M $tuned_params
grub_users $grub_users
grub_arg --unrestricted
grub_class centos
```

OK I assume the BLS config (and the extra stuff) gets created when the grub RPMs get installed or something that happens as part of the rpm transaction and now I understand why it needs to be fixed up; to remove all the osbuild related goup.

I haven't run into the need for `fix-bls` because we aren't doing an RPM transaction, but rather using OSTree which handles the BLS stuff for us.

What I don't really understand is how the `centos_9-s390x-qcow2-boot.json` works. It appears to be doing the same thing we are 

Oh wait..


I now am running the [s390x config](https://github.com/osbuild/osbuild-composer/blob/e289b763e7bbadb7a95e566d9d365efb79d47557/test/data/manifests/centos_9-s390x-qcow2-boot.json#L4729-L4734).

Before the `fix-bls` stage on s390x:


```
bash-5.1# cat /run/osbuild/tree/boot/loader/entries/ffffffffffffffffffffffffffffffff-5.14.0-200.el9.s390x.conf 
title  (5.14.0-200.el9.s390x) 
version 5.14.0-200.el9.s390x
linux /boot/vmlinuz-5.14.0-200.el9.s390x
initrd /boot/initramfs-5.14.0-200.el9.s390x.img
options root=UUID=6e4ff95f-f662-45ee-a82a-bdf44a2d0b75 console=tty0 console=ttyS0,115200n8 no_timer_check net.ifnames=0 crashkernel=1G-4G:192M,4G-64G:256M,64G-:512M
id -20240131005438-5.14.0-200.el9.s390x
grub_users $grub_users
grub_arg --unrestricted
grub_class kernel
```

and after it's the exact same all the way through to the `zipl.inst` stage.


So the reason it's working for the RHEL/CentOS s390x config you shared is that somehow the the BLS config that gets pre-populated already has a path that works even though it's actually invalid (if this were a GRUB system it wouldn't boot IIUC). 

> 
> One thing I'm not happy about here is (again) the stage detecting things about the tree or the mounts. If a `prefix` is needed for some of the options and the stage has a device with mounts attached to it, why does the prefix need to be detected? The value should be known when creating the manifest and the paths can be set accordingly in the stage options, can't they?

I guess we can do that, but we really shouldn't have to. Basically if we assume the config in the BLS entry is correct (i.e. the path to the kernel and initrd should be the path from the root of the filsystem in which the blscfg is on) then we wouldn't have to. We'd just be able to prepend the path to that mounted filesystem.

Reply:
another way of looking at this is there is a bug in `org.osbuild.fix-bls`:


```
[dustymabe@hattop ~]$ python3
Python 3.12.1 (main, Dec  8 2023, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import re
>>> path_re = re.compile(r"(/.*)+/boot")
>>> line='linux /root/var/osbuild/store/stage/uuid-e412a9ca6de34867a04f812cd0dcd623/data/tree/boot/vmlinuz-5.14.0-214.el9.x86_64'
>>> prefix=''
>>> print(path_re.sub(prefix, line))
linux /vmlinuz-5.14.0-214.el9.x86_64
>>> 
>>> line='linux /boot/vmlinuz-5.14.0-200.el9.s390x'
>>> print(path_re.sub(prefix, line))
linux /boot/vmlinuz-5.14.0-200.el9.s390x
```

Reply:
Right, the stage assumes that there are path components ahead of `/boot` that need to be dropped (so the regex starts with `(/.*)+` and the whole thing doesn't match a plain `/boot`).

Reply:
> OK I assume the BLS config (and the extra stuff) gets created when the grub RPMs get installed or something that happens as part of the rpm transaction and now I understand why it needs to be fixed up; to remove all the osbuild related goup.

Yup, exactly.

Reply:
> Right, the stage assumes that there are path components ahead of `/boot` that need to be dropped (so the regex starts with `(/.*)+` and the whole thing doesn't match a plain `/boot`).

Do you consider that to be a bug as I argue in https://github.com/osbuild/osbuild/pull/1558#discussion_r1472200040 ?

Reply:
So just to take a step back here and understand the change in general.
The zipl command args `--image` and `--ramdisk` take real paths to the kernel and initrd (the actual files).  The `find_kernel()` function reads the bls entries to find them.  And since it works, it means the bls entries are `/boot/vmlinuz...`.

Digging into this, it seems the bls entries do indeed include `/boot/vmlinuz-...` immediately after the rpm stage, so fix-bls on s390x is a no-op.

So yes, on an image built just now, with a separate boot partition, the bls looks like this:
```
# cat /mnt/loader/entries/ffffffffffffffffffffffffffffffff-6.6.8-200.fc39.s390x.conf
title  (6.6.8-200.fc39.s390x)
version 6.6.8-200.fc39.s390x
linux /boot/vmlinuz-6.6.8-200.fc39.s390x
initrd /boot/initramfs-6.6.8-200.fc39.s390x.img
options root=UUID=a4f79e2b-ce29-4e4f-b4b6-a9e4e66b07d0 ro no_timer_check console=ttyS0,115200n8 biosdevname=0 net.ifnames=0
id -20240131151940-6.6.8-200.fc39.s390x
grub_users $grub_users
grub_arg --unrestricted
grub_class kernel
```
(with `/boot`).

But then even on the machine I'm running on (a beaker), I see the same:
```
[root@ibm-z-530 images]# cat /boot/loader/entries/56af4fd613914413992e7d2dc1878319-6.6.13-200.fc39.s390x.conf
title  (6.6.13-200.fc39.s390x)
version 6.6.13-200.fc39.s390x
linux /boot/vmlinuz-6.6.13-200.fc39.s390x
initrd /boot/initramfs-6.6.13-200.fc39.s390x.img
options root=UUID=70417686-01d1-41c5-9afe-30aba17cdb6a rd.dasd=0.0.0120 rd.lvm.lv=fedora_ibm-z-530/root cio_ignore=all,!condev
id -20240131135953-6.6.13-200.fc39.s390x
grub_users $grub_users
grub_arg --unrestricted
grub_class kernel
[root@ibm-z-530 images]# findmnt /boot
TARGET
      SOURCE      FSTYPE OPTIONS
/boot /dev/dasda1 xfs    rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,noquota
```
:thinking: 

What am I missing?

Reply:
> > Right, the stage assumes that there are path components ahead of `/boot` that need to be dropped (so the regex starts with `(/.*)+` and the whole thing doesn't match a plain `/boot`).
> 
> Do you consider that to be a bug as I argue in [#1558 (comment)](https://github.com/osbuild/osbuild/pull/1558#discussion_r1472200040) ?

I guess it depends.  I'm certain the original regex was written for a specific purpose (the paths getting written by grub config and including the osbuild paths) and wasn't meant to be too general, so in that respect no.  But if we need to cover other use cases, like BLS entries that don't include extra paths but need to be changed/fixed to drop `/boot`, then maybe we should change it.
Still curious why/how it works though.

Reply:
>And since it works, it means the bls entries are probably /boot/vmlinuz....

I've checked the template for Centos  and bls configs on RHEL9.s390x and F39.s390x:
```
$  cat /boot/loader/entries/5bfa965ae7de44adba93c94e8709ff8c-5.14.0-162.6.1.el9_1.s390x.conf                                                                                                                             
title  (5.14.0-162.6.1.el9_1.s390x)               
version 5.14.0-162.6.1.el9_1.s390x
linux /boot/vmlinuz-5.14.0-162.6.1.el9_1.s390x
initrd /boot/initramfs-5.14.0-162.6.1.el9_1.s390x.img 
```

```
            "mounts": [
              {
                "name": "root",
                "type": "org.osbuild.xfs",
                "source": "root",
                "target": "/"
              },
              {
                "name": "boot",
                "type": "org.osbuild.xfs",
                "source": "boot",
                "target": "/boot"
              }
```

`find_kernel` in this case returns proper paths with `/boot/`, therefore `zipl` get proper paths to kernel/ramdsik. 

CoreOs on the other hand would have only relative paths `ostree/....`, that's why we have to prepend `/boot`

Reply:
all - @achilleas-k and I met yesterday and did a deep dive on this (sorry for not updating the ticket at the time, I got pulled away for something else). I opened a new issue to document the current status and the suggested path forward: https://github.com/osbuild/osbuild/issues/1566

Please add any clarifying statements there if anything we found isn't accurate!